### PR TITLE
Compress history backup directory

### PR DESCRIPTION
**Pull request by daehn**
## What's new in this PR?

* Move `ZipArchive` from the UI project to sync-engine.
* Compress history backup directory.

### This PR depends on:

- [x] A new release of `wire-ios-data-model` after https://github.com/wireapp/wire-ios-data-model/pull/457 has been merged.

### CODE REVIEW

File: Source/SessionManager/SessionManager+Backup.swift [hunk 1]
```diff
@@ -18,14 +18,61 @@
...
     public func backupActiveAccount(completion: @escaping ((Result<URL>) -> ())) {
-        guard let userId = self.accountManager.selectedAccount?.userIdentifier else { return completion(.failure(BackupError.noActiveAccount)) }
-        StorageStack.backupLocalStorage(accountIdentifier: userId, applicationContainer: sharedContainerURL, completion: completion)
+        guard let userId = accountManager.selectedAccount?.userIdentifier,
+              let context = activeUserSession?.managedObjectContext,
+              let clientId = ZMUser(remoteID: userId, createIfNeeded: false, in: context)?.selfClient()?.remoteIdentifier
+        else { return completion(.failure(BackupError.noActiveAccount)) }
+        
+        StorageStack.backupLocalStorage(accountIdentifier: userId, clientIdentifier: clientId, applicationContainer: sharedContainerURL) { result in
+            completion(result.map(SessionManager.compress))
```
Comment (vytis) [inline Source/SessionManager/SessionManager+Backup.swift]: I think this is all happening on main thread, might take a long time if db is large

File: Source/SessionManager/SessionManager+Backup.swift [hunk 2]
```diff
@@ -18,14 +18,61 @@
...
+    private static func compressedBackupURL(for backup: StorageStack.BackupInfo) -> URL {
+        return backup.url.deletingLastPathComponent().appendingPathComponent(backup.metadata.backupFilename)
+    }
+}
+
+// MARK: - Compressed Filename
+
+fileprivate extension BackupMetadata {
+    
+    private static let fileExtension = "wirebackup"
```
Comment (vytis) [inline Source/SessionManager/SessionManager+Backup.swift]: probably should have `ios` in there somewhere

**Comment by daehn** [comment]
@vytis updated!

File: Source/SessionManager/SessionManager+Backup.swift [hunk 3]
```diff
@@ -18,14 +18,84 @@
...
+              let context = activeUserSession?.managedObjectContext,
+              let clientId = ZMUser.selfUser(in: context).selfClient()?.remoteIdentifier
+        else { return completion(.failure(BackupError.noActiveAccount)) }
+        
+        StorageStack.backupLocalStorage(
+            accountIdentifier: userId,
+            clientIdentifier: clientId,
+            applicationContainer: sharedContainerURL,
+            dispatchGroup: dispatchGroup,
+            completion: { [weak self] in SessionManager.handle(result: $0, dispatchGroup: self?.dispatchGroup, completion: completion) }
```
Comment (vytis) [inline Source/SessionManager/SessionManager+Backup.swift]: Futures would already be quite good here :)

File: Source/SessionManager/SessionManager+Backup.swift [hunk 4]
```diff
@@ -18,14 +18,84 @@
...
+
     enum BackupError: Error {
         case noActiveAccount
+        case compressionError
+    }
+
+    public func backupActiveAccount(completion: @escaping BackupResultClosure) {
+        guard let userId = accountManager.selectedAccount?.userIdentifier,
+              let context = activeUserSession?.managedObjectContext,
+              let clientId = ZMUser.selfUser(in: context).selfClient()?.remoteIdentifier
```
Comment (typfel) [inline Source/SessionManager/SessionManager+Backup.swift]: You can get the selfClient directly from the user session `activeUserSession.selfUserClient`

File: Source/SessionManager/SessionManager+Backup.swift (hunk 3 above)
Comment (daehn) [reply Source/SessionManager/SessionManager+Backup.swift]: Yeah, was thinking that as well 👍

**Pull request merged**